themes/custom/az_barrio/az_barrio.theme

```diff
@@ -396,31 +397,38 @@ function az_barrio_primary_logo() {
 
   // Set primary logo.
   if (!empty($primary_logo_path)) {
-    $url = file_create_url($primary_logo_path['path']);
 
     // Inline SVG logo.
     if (theme_get_setting('az_barrio_logo_svg_inline') === TRUE && $primary_logo_path_parts['extension'] === 'svg') {
       $svg = file_get_contents(DRUPAL_ROOT . '/' . $primary_logo_path['url']);
       $rendered_image = render($svg);
       $image_markup = Markup::create($rendered_image);
+      $primary_logo_link_classes[] = 'qs-logo-svg-inline';
       $return = Link::createFromRoute($image_markup, '<front>', [], [
         'attributes' => [
           'title' => $primary_title_text,
-          'class' => ['header_logo_svg'],
+          'class' => $primary_logo_link_classes,
           'rel' => 'home',
         ],
       ]
       );
     }
     else {
-      $image = '<img src="' . $url . '" alt="' . $primary_alt_text . '" class="img-fluid" />';
-      $rendered_image = render($image);
-      $image_markup = Markup::create($rendered_image);
-      $return = Link::createFromRoute($image_markup, '<front>', [], [
+      $image_renderable = [
+        '#theme' => 'image',
+        '#uri' => $primary_logo_path['path'],
+        '#alt' => $primary_alt_text,
+        '#attributes' => [
+          'class' => ['img-fluid'],
+        ]
+      ];
+      $renderer = \Drupal::service('renderer');
+      $renderer->addCacheableDependency($image_renderable, $config);
+      $image_rendered = $renderer->render($image_renderable);
```

Comment:
I don't think this `$renderer->render()` step is actually necessary because the `$text` parameter for [`Link::createFromRoute()` ](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Link.php/function/Link%3A%3AcreateFromRoute/9.1.x)can either be a string or a render array.

Reply:
OK so I suppose the question is for `$text`
Do I pass the 

1. `$image_renderable` <- won't this not add the cache tags?
2. `$image_rendered` 
3. `$renderer`

Reply:
I think if you just pass `$image_renderable`, everything will continue to work including the cache tags.

Reply:
Sounds like magic. haha

Reply:
I'm surprised that this is not a part of schema or config in some way if it just works magically like that.

Reply:
This is where some of the magic happens: https://git.drupalcode.org/project/drupal/-/blob/9.1.x/core/lib/Drupal/Core/Utility/LinkGenerator.php#L89-91

Reply:
So services are kind of like the hook list of old, where you add stuff to the service, and then it is called in a specified order?

Reply:
I'll just read this https://symfony.com/doc/3.4/service_container.html